src/recipe_parser.zig

```diff
@@ -0,0 +1,264 @@
+			allocator.free(combo);
+		}
+		inputCombos.deinit();
+	}
+	var keyList: main.List(std.StringHashMap([]const u8)) = .initCapacity(allocator, startingParsedItems.items.len);
+	defer {
+		for(keyList.items) |*keys| {
+			keys.deinit();
+		}
+		keyList.deinit();
+	}
+	for(startingParsedItems.items) |item| {
+		const inputs = allocator.alloc(ItemStack, recipe.len);
+		inputs[0] = item.item;
+		inputCombos.append(inputs);
+		keyList.append(item.keys);
+	}
+	while(remainingItems.len > 1) {
+		remainingItems = remainingItems[1..];
+		const startIndex = inputCombos.items[0].len - remainingItems.len;
+		var newKeyList: main.List(std.StringHashMap([]const u8)) = .init(allocator);
+		var newInputCombos: main.List([]ItemStack) = .init(allocator);
+
+		for(keyList.items, inputCombos.items) |*keys, inputs| {
+			const parsedItems = try parseRecipeItem(allocator, remainingItems[0], keys);
+			defer parsedItems.deinit();
+			for(parsedItems.items) |item| {
+				const newInputs = allocator.dupe(ItemStack, inputs);
+				newInputs[startIndex] = item.item;
+				newInputCombos.append(newInputs);
+				newKeyList.append(item.keys);
+			}
+		}
+
+		for(keyList.items) |*keys| {
+			keys.deinit();
+		}
+		keyList.deinit();
+		keyList = newKeyList;
+		for(inputCombos.items) |combo| {
+			allocator.free(combo);
+		}
+		inputCombos.deinit();
+		inputCombos = newInputCombos;
+	}
+	return inputCombos;
+}
+
+pub fn parseRecipe(allocator: NeverFailingAllocator, zon: ZonElement, list: *main.List(Recipe)) !void {
+	const inputs = zon.getChild("inputs").toSlice();
+	const recipeItems = allocator.alloc(ZonElement, inputs.len + 1);
+	@memmove(recipeItems[0..inputs.len], inputs);
+	recipeItems[inputs.len] = zon.getChild("output");
+	defer allocator.free(recipeItems);
+
+	const itemCombos = try generateItemCombos(allocator, recipeItems);
+	defer itemCombos.deinit();
+	for(itemCombos.items) |itemCombo| {
+		defer allocator.free(itemCombo);
+		const parsedInputs = itemCombo[0 .. itemCombo.len - 1];
+		const output = itemCombo[itemCombo.len - 1];
+		const recipe = Recipe{
+			.sourceItems = list.allocator.alloc(BaseItemIndex, parsedInputs.len),
+			.sourceAmounts = list.allocator.alloc(u16, parsedInputs.len),
+			.resultItem = output.item.?.baseItem,
+			.resultAmount = output.amount,
+		};
+		for(parsedInputs, 0..) |input, i| {
+			recipe.sourceItems[i] = input.item.?.baseItem;
+			recipe.sourceAmounts[i] = input.amount;
+		}
+		list.append(recipe);
+		if(zon.get(bool, "reversible", false)) {
+			if(recipe.sourceItems.len == 1) {
+				var reversedRecipe = Recipe{
+					.sourceItems = list.allocator.alloc(BaseItemIndex, 1),
+					.sourceAmounts = list.allocator.alloc(u16, 1),
+					.resultItem = recipe.sourceItems[0],
+					.resultAmount = recipe.sourceAmounts[0],
+				};
+				reversedRecipe.sourceItems[0] = recipe.resultItem;
+				reversedRecipe.sourceAmounts[0] = recipe.resultAmount;
+				list.append(reversedRecipe);
+			} else {
+				return error.InvalidReversibleRecipe;
```

Comment:
Generally it's more readable to have the error path first, with an else branch it is harder to see what the actual condition is.
```zig
if(recipe.sourceItems.len == 1) {
	...
} else {
	return error.InvalidReversibleRecipe;
}
```
↓
```zig
if(recipe.sourceItems.len != 1) {
	return error.InvalidReversibleRecipe;
}
...
```